waspc/src/Wasp/Generator/Monad.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.Generator.Monad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    catchGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Arrow (left)
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+import Data.List.NonEmpty (NonEmpty, fromList)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the error that halted generation. Every error is fatal.
+-- There may optionally be additional errors or non-fatal warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError = GenericGeneratorError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+
+data GeneratorWarning = GenericGeneratorWarning String
+
+instance Show GeneratorWarning where
+  show (GenericGeneratorWarning e) = e
+
+-- Runs the generator and either returns a result, or a list of 1 or more errors.
+-- Even if successful there may be warnings, so they are always included.
+runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
+runGenerator generator =
+  let (result, state') = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
+   in (warnings state', left (\_ -> fromList $ errors state') result)
+  where
+    state = GeneratorState {warnings = [], errors = []}
+
+-- This logs a warning but does not short circuit the computation.
+logGeneratorWarning :: GeneratorWarning -> Generator ()
+logGeneratorWarning w = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
+  GeneratorState {errors = errors', warnings = w : warnings'}
+
+-- This logs an error and does throw, thus short-circuiting the computation unless caught.
```

Comment:
```suggestion
-- This logs an error and does throw, thus short-circuiting the computation until caught.
```
I think `until` might be more correct, because if you have g1 call g2 that calls g3, and g3 throws, but g1 catches, you will short-circuit g3 and g2, but not g1. So it short-circuited until g1 caught it. While `unless` suggests that no short-circuiting happens at all, but it does in g2 and g3. Ok this is extreme nitpicking and possibly searching for meaning where it shouldn't be but I still thought it might be useful to be explicit.

Reply:
Yep, no that is more accurate and worth a swap. `unless` gives the wrong impression for sure!